### PR TITLE
[JENKINS-46759] Fixed bug in build queue filtering for views

### PR DESCRIPTION
**Pull request by jonsten**
Fixed so that pipeline build parts show up in the build queue of views which filter build queue.
I've created a test case but I'm not 100% sure that this is the best way to test it, feedback is very much appreciated!

See [JENKINS-46759](https://issues.jenkins-ci.org/browse/JENKINS-46759).

### Proposed changelog entries

* Fixed so that pipeline build parts show up in the build queue of views which filter build queue

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

### CODE REVIEW

**Review by oleg-nenashev — APPROVED** [review]
I am not sure about the performance impact, but LGTM from the behavior PoV.

**Comment by oleg-nenashev** [comment]
PR build failure is unrelated, retriggering the CI

**Comment by jonsten** [comment]
Thank you for the approval Oleg! Regarding performance impacts, looking at the current pipeline implementation it seems like it comes down to how expensive it is to evaluate `StepContext.get(Run.class)`!

File: core/src/main/java/hudson/model/View.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/View.java]: I wonder how likely a bug is that results in an infinite loop here. Still a bug elsewhere (`this.getOwnerTask() == this`, but I think we should break out after 10 or so iterations and log a warning.

File: core/src/main/java/hudson/model/View.java
Comment (jglick) [inline core/src/main/java/hudson/model/View.java]: Deliberately getting rid of the `else` and thus running the `if (qi.task instanceof AbstractProject<?, ?>)` block even if we added to `result` in the loop here? That seems suspicious. If I read this correctly, for a typical `FreestyleProject` you are going to add two entries.

Maybe you meant to add `SOMELABEL:` to the `for`-loop and `continue` it, rather than `break`ing out of the `do` loop and falling through to the `AbstractProject` check.

File: core/src/main/java/hudson/model/View.java
Comment (jglick) [reply core/src/main/java/hudson/model/View.java]: I see no reason for concern. If you override `getOwnerTask` you obviously need to ensure that the chain terminates.

IIUC this loop could be written as

```java
Queue.Task currentTask = ci.task;
while (true) {
    if (items.contains(currentTask)) {
        result.add(qi);
        break;
    }
    Queue.Task next = currentTask.getOwnerTask();
    if (next == currentTask) {
        break;
    } else {
        currentTask = next;
    }
}
```

which may or may not be easier to grasp.

File: core/src/test/java/hudson/model/ViewTest.java
Comment (jglick) [inline core/src/test/java/hudson/model/ViewTest.java]: Fine although I generally do not trust Mockito/PowerMock tests at all; you anyway need to go back and confirm the actual behavior (in this case using a downstream PR in `workflow-durable-task-step-plugin`).

**Review by jglick — COMMENTED** [review]
Looks like a possible logic bug here. Otherwise sounds reasonable.

File: core/src/main/java/hudson/model/View.java
Comment (jonsten) [reply core/src/main/java/hudson/model/View.java]: You are correct, good catch! A label crossed my mind but didn't know it was kosher to use in Jenkins code base! I will take this and your other comment into consideration and try to redesign the loop. Thanks!

File: core/src/main/java/hudson/model/View.java
Comment (jonsten) [reply core/src/main/java/hudson/model/View.java]: As I said in my other comment, the logic is pretty complex, I will try to redesign!

File: core/src/test/java/hudson/model/ViewTest.java
Comment (jonsten) [reply core/src/test/java/hudson/model/ViewTest.java]: Okay, will do! Is it ok to add an actual Jenkins instance test in the downstream plugin? Sure feels more robust that way!

File: core/src/main/java/hudson/model/View.java
Comment (daniel-beck) [reply core/src/main/java/hudson/model/View.java]: > obviously

Bug report coming soon to a `robustness` label near you 😉

Comment (jonsten) [reply core/src/main/java/hudson/model/View.java]: BTW @daniel-beck how do I log a warning, create static Logger object in View class and then log to warning on that object? Or is there another way?

**Comment by jonsten** [comment]
So what is the workflow here, do I need to notify anyone for a new review or just wait? Cause I can't remove the needs-fix label...

**Review by oleg-nenashev — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
@jonsten My plan it to merge it towards next weekly release if there is no negative feedback from @jenkinsci/code-reviewers

**Comment by jonsten** [comment]
Okay, great, thanks!

**Comment by daniel-beck** [comment]
Haven't looked closely enough to approve, but no API addition to non-test code, so if the downstream PR (still TODO AFAICT) reveals problems, those can be resolved then.

**Comment by jonsten** [comment]
@daniel-beck Regarding downstream PR and test, as I see it I will have to add it after this PR has been merged and release, i.e. the downstream test will fail until that plugin's test dependency can be updated to the core version where this is fixed. Or is there another procedur for handling situations like this?

**Comment by daniel-beck** [comment]
@jonsten 

> Or is there another procedur for handling situations like this?

Anyone with an account is allowed to upload snapshots to our Maven repository. So `mvn deploy` a Jenkins core snapshot, and use a timestamped Maven dependency (not just `-SNAPSHOT`) in the downstream PR.

But note that I am saying this is not actually necessary since there are no API additions.

**Comment by jonsten** [comment]
Thanks @daniel-beck , unfortunately that is outside of my maven skills, but as you said, no downstream test is necessary.

Please correct me if I'm wrong, there isn't anything holding this PR back?

**Comment by daniel-beck** [comment]
I suspect @oleg-nenashev misread my comment to mean the opposite. Anyway, summoning him to clarify.

**Review by oleg-nenashev — APPROVED** [review]
I am not 100% sure `10` is enough for `FILTER_LOOP_MAX_COUNT`, but it is definitely better than the original behavior. Hence 👍 

@daniel-beck sorry, I have missed this PR and failed to remember which comment you meant. Just reviewed from scratch

**Comment by jonsten** [comment]
@oleg-nenashev 10 in look count should be fine for all cases that I know of right now. AFAIK pipelines is the only job type which needs this and it reach the root in the second iteration. It's not like we have to walk up the entire step graph to find the root.

**Comment by oleg-nenashev** [comment]
@jglick @daniel-beck are you fine with merge?

**Comment by daniel-beck** [comment]
@oleg-nenashev Unchanged since

> Haven't looked closely enough to approve, but no API addition to non-test code, so if the downstream PR (still TODO AFAICT) reveals problems, those can be resolved then.

**Comment by oleg-nenashev** [comment]
🚢 🇮🇹 then

**Comment by jonsten** [comment]
Great! Thanks for the help everyone :)

**Pull request merged**
